src/parcsr_mv/par_csr_matop.c

```diff
@@ -5265,125 +5155,108 @@ hypre_ParcsrAdd( HYPRE_Complex        alpha,
-         }
+         ns = ii*size+rest;
+         ne = (ii+1)*size+rest;
       }
 
-      C_offd_i[ii+1] = nnz_offd_C;
-   }
-   C_offd_i[num_rows_diag_A] = nnz_offd_C;
+      marker_diag = hypre_TAlloc(HYPRE_Int, num_cols_diag_A, HYPRE_MEMORY_HOST);
+      hypre_CSRMatrixAddFirstPass(ns, ne, twspace, marker_diag,
+                                  NULL, NULL, A_diag, B_diag,
+                                  num_rows_diag_C, num_rownnz_diag_C,
+                                  num_cols_diag_C, rownnz_diag_C,
+                                  memory_location_C, C_diag_i, &C_diag);
+      hypre_CSRMatrixAddSecondPass(ns, ne, twspace, marker_diag,
+                                   NULL, NULL, rownnz_diag_C,
+                                   alpha, beta, A_diag, B_diag, C_diag);
+      hypre_TFree(marker_diag, HYPRE_MEMORY_HOST);
 
-   /* Correct C_offd_i */
-   if (rownnz_offd_C)
-   {
-      for (i = 0; i < num_rownnz_offd_C - 1; i++)
+     /*-----------------------------------------------------------------------
+      *  Compute C_offd = alpha*A_offd + beta*B_offd
+      *-----------------------------------------------------------------------*/
+
+      size = num_rownnz_offd_C/num_threads;
+      rest = num_rownnz_offd_C - size*num_threads;
+      if (ii < rest)
       {
-         for (ii = rownnz_offd_C[i] + 1; ii < rownnz_offd_C[i+1]; ii++)
-         {
-            C_offd_i[ii+1] = C_offd_i[rownnz_offd_C[i]+1];
-         }
+         ns = ii*size+ii;
+         ne = (ii+1)*size+ii+1;
       }
-      if (num_rownnz_offd_C)
+      else
       {
-         for (ii = rownnz_offd_C[num_rownnz_offd_C - 1] + 1; ii < num_rows_diag_A; ii++)
-         {
-            C_offd_i[ii+1] = C_offd_i[rownnz_offd_C[num_rownnz_offd_C - 1]+1];
-         }
+         ns = ii*size+rest;
+         ne = (ii+1)*size+rest;
       }
-   }
 
+      marker_offd = hypre_TAlloc(HYPRE_Int, num_cols_offd_C, HYPRE_MEMORY_HOST);
+      hypre_CSRMatrixAddFirstPass(ns, ne, twspace, marker_offd,
+                                  A2C_offd, B2C_offd, A_offd, B_offd,
+                                  num_rows_offd_C, num_rownnz_offd_C,
+                                  num_cols_offd_C, rownnz_offd_C,
+                                  memory_location_C, C_offd_i, &C_offd);
+      hypre_CSRMatrixAddSecondPass(ns, ne, twspace, marker_offd,
+                                   A2C_offd, B2C_offd, rownnz_offd_C,
+                                   alpha, beta, A_offd, B_offd, C_offd);
+      hypre_TFree(marker_offd, HYPRE_MEMORY_HOST);
+   } /* end of omp parallel region */
+
+   /* Free memory */
+   hypre_TFree(twspace, HYPRE_MEMORY_HOST);
+   hypre_TFree(A2C_offd, HYPRE_MEMORY_HOST);
+   hypre_TFree(B2C_offd, HYPRE_MEMORY_HOST);
+
+   /* Create ParCSRMatrix C */
    row_starts_C = hypre_TAlloc(HYPRE_BigInt, 2, HYPRE_MEMORY_HOST);
    col_starts_C = hypre_TAlloc(HYPRE_BigInt, 2, HYPRE_MEMORY_HOST);
    memcpy(row_starts_C, hypre_ParCSRMatrixRowStarts(A), 2*sizeof(HYPRE_BigInt));
```

Comment:
Maybe better use the one in hypre's memory model: `hypre_TMemcpy` Host To Host. It has the same effect for now.

Reply:
Yes, agreed! I realized memcpy is used in other parts of the code as well. Maybe we need to update those